### PR TITLE
ESP32/ESP32-C3: Fix leak of semaphores created by Wi-Fi kernel thread

### PR DESCRIPTION
**Pull request by gustavonihei**
## Summary
This PR intends to provide an alternative solution to the fix provide on https://github.com/apache/incubator-nuttx/pull/4092 for **ESP32-C3**, and extend it to the **ESP32** Wi-Fi driver.

Originally, the semaphores created by the Wi-Fi kernel thread for handling application connection requests were not being destroyed, causing the a memory leak of 32 bytes for every `wapi` invocation.

The **ESP32-C3** solution is being refactored in order to avoid the usage of `on_exit` in driver code, see: https://github.com/apache/incubator-nuttx/pull/6197

## Impact
Fix for semaphore leak on **ESP32** and refactor of current solution on **ESP32-C3**.

## Testing
`esp32-devkitc:wapi`
`esp32c3-devkit:wapi`

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by pkarashchenko — APPROVED** [review]
LGTM!

**Comment by pkarashchenko** [comment]
One question from my side. Do we expect multi-instance of wifi driver? If not, then maybe we can just have a global semaphore pointer and use that on/atexit?
I mean I do not fully understand neither pthread_key nor task_tls_ usage here.
Maybe it is reasonable to have static pointer to semaphore and if it is NULL, the allocate and init it. Then at/onexit the static pointer can be checked again and free the memory. Am I missing something?

File: arch/risc-v/src/esp32c3/esp32c3_wifi_adapter.c [hunk 1]
```diff
@@ -1338,41 +1334,40 @@ static int IRAM_ATTR wifi_is_in_isr(void)
...
-          break;
+          wlerr("Failed to create task local key\n");
+          return NULL;
         }
+
+      wifi_task_key = ret;
     }
 
-  if (i >= CONFIG_SCHED_EXIT_MAX)
+  sem = (void *)task_tls_get_value(wifi_task_key);
```
Comment (pkarashchenko) [inline arch/risc-v/src/esp32c3/esp32c3_wifi_adapter.c:1353]: @xiaoxiang781216 do you think we need to add
```
info->ta_telem[candidate] = 0;
```
after `dtor((void *)elem);` in `task_tls_destruct()`?

I'm just thinking that if the kthread exits and then started for the second time, then `wifi_task_key` will be non-null, but `task_tls_get_value(wifi_task_key);` will return non-null pointer that is already freed by `esp_thread_semphr_free`.

**Comment by gustavonihei** [comment]
The Wi-Fi library creates a new semaphore for every thread that performs connection operations, so we cannot have a global pointer.
Since these semaphores are thread-local, this motivated the initial implementation based on `pthread_key_t`, so that the semaphores were being stored in Thread Local Storage, and then could be destroyed on thread termination. The solution here was based on the one implemented for ESP-IDF, which works as expected.

But on NuttX it resulted in the semaphores not being destroyed. I'll try to explain why.
The Wi-Fi library operates in a dedicated Kernel Thread, named `wifi`. But the `pthread_key_t` and the destructor for the semaphores were allocated to the Thread Local Storage of the `init` thread.
Every network-related request from the application will be handled by the `wifi` kernel thread and its child pthreads. The issue is that those child pthreads do not belong to the same Task Group from the `init` thread, which is the one whose TLS area contains the semaphore destructor.

So the catch here is that NuttX provides this process-like abstraction which segregates pthreads created from different tasks. So a pthread created from **Task B** won't be able to share keys with another pthread from **Task A**.

**Comment by gustavonihei** [comment]
> Do we expect multi-instance of wifi driver?

No, the Wi-Fi driver does not expect to be initialized more than once.

File: arch/risc-v/src/esp32c3/esp32c3_wifi_adapter.c (hunk 1 above)
Comment (gustavonihei) [reply arch/risc-v/src/esp32c3/esp32c3_wifi_adapter.c:1353]: > I'm just thinking that if the kthread exits and then started for the second time, then `wifi_task_key` will be non-null, but `task_tls_get_value(wifi_task_key);` will return non-null pointer that is already freed by `esp_thread_semphr_free`.

In this proposed scenario where a given kernel thread is started a second time, `task_tls_get_value` will index a newly created `task_info_s` struct. So even though the key value has not been reset due to static storage duration, the `task_tls_get_value` couldn't do much for preventing this wrong usage of the `tlsindex` key.

**Comment by yamt** [comment]
my config had TLS_TASK_NELEM=0 and i got a bit surprised by this change breaking the build.
does Kconfig have a way to solve this kind of things in a bit more user friendly way?

**Review by pkarashchenko — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
Task tls is disabled by default.:(

**Comment by pkarashchenko** [comment]
> > One question from my side. Do we expect multi-instance of wifi driver? If not, then maybe we can just have a global semaphore pointer and use that on/atexit?
> 
> The Wi-Fi library creates a new semaphore for every thread that performs connection operations, so we cannot have a global pointer. Since these semaphores are thread-local, this motivated the initial implementation based on `pthread_key_t`, so that the semaphores were being stored in Thread Local Storage, and then could be destroyed on thread termination. The solution here was based on the one implemented for ESP-IDF, which works as expected.
> 
> But on NuttX it resulted in the semaphores not being destroyed. I'll try to explain why. The Wi-Fi library operates in a dedicated Kernel Thread, named `wifi`. But the `pthread_key_t` and the destructor for the semaphores were allocated to the Thread Local Storage of the `init` thread. Every network-related request from the application will be handled by the `wifi` kernel thread and its child pthreads. The issue is that those child pthreads do not belong to the same Task Group from the `init` thread, which is the one whose TLS area contains the semaphore destructor.
> 
> So the catch here is that NuttX provides this process-like abstraction which segregates pthreads created from different tasks. So a pthread created from **Task B** won't be able to share keys of type `pthread_key_t` with another pthread from **Task A**.

Yes. I just reexamined the task tls code and see that I did a wrong assumption.

**Pull request merged**